mlir/lib/Catalyst/Transforms/QnodeToAsyncPatterns.cpp

```diff
@@ -0,0 +1,119 @@
+
+#include "mlir/Dialect/Async/IR/Async.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+namespace catalyst {
+
+struct CallOpToAsyncOPRewritePattern : public mlir::OpRewritePattern<func::CallOp> {
+    using mlir::OpRewritePattern<func::CallOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(func::CallOp op,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        SymbolRefAttr symbol = dyn_cast_if_present<SymbolRefAttr>(op.getCallableForCallee());
+        func::FuncOp func = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, symbol);
+        if (!func) {
+            // Nothing to change.
+            return failure();
+        }
+
+        // Check for Call ops that have QNode func ops
+        if (!func->hasAttrOfType<UnitAttr>("qnode")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        if (op->hasAttrOfType<UnitAttr>("transformed")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        TypeRange retTy = op.getResultTypes();
+        SmallVector<Value> dependencies; /* = empty */
+        SmallVector<Value> operands;     /* = empty */
+        auto noopExec = [&](OpBuilder &executeBuilder, Location executeLoc,
+                            ValueRange executeArgs) {};
+
+        rewriter.updateRootInPlace(op, [&] { op->setAttr("transformed", rewriter.getUnitAttr()); });
+        IRMapping map;
+        auto executeOp =
+            rewriter.create<async::ExecuteOp>(op.getLoc(), retTy, dependencies, operands, noopExec);
+        {
+            PatternRewriter::InsertionGuard insertGuard(rewriter);
+            rewriter.setInsertionPoint(executeOp.getBody(), executeOp.getBody()->end());
+            Operation *cloneOp = op->clone(map);
+            rewriter.insert(cloneOp);
+            rewriter.create<async::YieldOp>(op.getLoc(), cloneOp->getResults());
+        }
+
+        auto asyncValues = executeOp.getResults();
+
+        // Only one use
+        // And in the same basic block
+        // If it is not in the same basic block
+        // It may be in a loop.
+        // And therefore we may execute the drop_ref multiple times
+        // which is a runtime error.
+        auto results = op.getResults();
+        if (results.size() == 1) {
+            for (auto result : results) {
+                if (result.hasOneUse()) {
+                    for (Operation *user : result.getUsers()) {
+                        if (op->getBlock() == user->getBlock()) {
+                            rewriter.setInsertionPoint(user);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
This nest is a bit hard to decipher at first glance, and the comments are super cryptic 😅 

My best guest is it jumps ahead to the first use of the callOp results, but why only when there is exactly one result? And why only when that result has one use?

In either case a comment with a clear, high-level description of the step being taken would help :)

Reply:
I have updated the comment, and added a todo to come with a better algorithm for the delay of await ops.